clicommand/pipeline_upload.go

```diff
@@ -802,28 +811,81 @@ stepsLoop:
 			ica.changedPaths = cps
 		}
 
-		// Parse and test the glob pattern against the paths.
-		pattern, ok := ic.(string)
-		if !ok {
-			l.Warn("if_changed value must be a string containing a glob pattern (was a %T)\n"+
-				"The step will not be skipped.", ic)
+		// Parse and test the glob pattern(s) against the paths.
+		include := ifChangedPatterns(l, ifChangedValue, "if_changed")
+		if len(include) == 0 {
 			continue
 		}
-
-		glob, err := zzglob.Parse(pattern)
-		if err != nil {
-			l.Warn("if_changed value %q couldn't be parsed as a glob pattern: %v\n"+
-				"The step will not be skipped.", pattern, err)
-			continue
+		exclude := ifChangedPatterns(l, ifChangedExcludeValue, "if_changed_exclude")
+
+		changedPaths := ica.changedPaths
+		if len(exclude) > 0 {
+			// First, remove paths that are excluded.
+			changedPaths = slices.DeleteFunc(slices.Clone(ica.changedPaths), func(cp string) bool {
+				for _, g := range exclude {
+					if g.Match(cp) {
+						return true
+					}
+				}
+				return false
+			})
 		}
-		for _, cp := range ica.changedPaths {
-			if glob.Match(cp) {
-				continue stepsLoop // one of the paths changed
+
+		// For each remaining changed path, test it against each pattern in `include`.
+		// If it's included, and wasn't excluded, we _don't skip_ this step. Got it?
+		for _, cp := range changedPaths {
+			for _, g := range include {
+				if g.Match(cp) {
+					continue stepsLoop
+				}
 			}
 		}
 
-		// Glob matched no changed file paths - this step is now skipped.
+		// The globs matched no changed file paths - this step is now skipped.
 		// Note that the "skip" string is limited to 70 characters.
 		content["skip"] = ifChangedSkippedMsg
 	}
 }
+
+// ifChangedPatterns converts a value for `if_changed` or `if_changed_exclude`
+// into a list of parsed globs.
+func ifChangedPatterns(l logger.Logger, value any, name string) []*zzglob.Pattern {
+	if value == nil {
+		return nil
+	}
+
+	var patterns []string
+	switch x := value.(type) {
+	case string:
+		patterns = []string{x}
+
+	case []any:
+		for i, item := range x {
+			patt, ok := item.(string)
+			if !ok {
+				l.Warn("Each %s value must be a string containing a glob pattern, but the item at index %d was a %T\n"+
+					"The step will not be skipped.", name, i, item)
+				return nil
```

Comment:
what does `nil` result in? does it skip the step or run it?

Reply:
`nil` - it runs the step. This particular code path would happen if someone wrote:

```yaml
if_changed:
  - something
  - null
  - something/else
```